addons/promote/common/src/main/java/org/commonjava/indy/promote/data/PromotionManager.java

```diff
@@ -519,189 +501,27 @@ public PathsPromoteResult promotePaths( final PathsPromoteRequest request, final
-                {
-                    String msg =
-                            "Path promotion failed. Target: " + request.getTarget() + ", Source: " + request.getSource()
-                                    + ", Reason: " + getStackTrace( ex );
-                    logger.warn( msg );
-                    ret = new PathsPromoteResult( request, msg );
-                }
-
-                if ( ret.getRequest().getCallback() != null )
-                {
-                    return callbackHelper.callback( ret.getRequest().getCallback(), ret );
-                }
-
-                return ret;
-            } )
-        );
-    }
-
-    /**
-     * Attempt to resumePathsPromote from a previously failing {@link PathsPromoteResult}. This is meant to handle cases where a transient (or correctable) error
-     * occurs on the server side, and promotion can proceed afterward. It works much like the {@link #promotePaths(PathsPromoteRequest, String)} call, using the pending
-     * paths list from the input result as the list of paths to process. The output {@link PathsPromoteResult} contains all previous completed paths PLUS
-     * any additional completed transfers when it is returned, thus providing a cumulative result to the user.
-     *
-     * @param result The result to resumePathsPromote
-     *
-     * @return The same result, with any successful path promotions moved from the pending to completed paths list, and the error cleared (or set to a
-     * new error)
-     *
-     * @throws PromotionException
-     * @throws IndyWorkflowException
-     */
-    public PathsPromoteResult resumePathsPromote( final PathsPromoteResult result, final String baseUrl )
-            throws PromotionException, IndyWorkflowException
-    {
-        final PathsPromoteRequest request = result.getRequest();
-
-        Future<PathsPromoteResult> future = submitResumePathsPromote( result, baseUrl );
-        if ( request.isAsync() )
-        {
-            return new PathsPromoteResult( request ).accepted();
-        }
-        else
-        {
-            try
-            {
-                return future.get();
-            }
-            catch ( InterruptedException | ExecutionException e )
-            {
-                logger.error( "Path promotion resume failed: " + request.getSource() + " -> " + request.getTargetKey(), e );
-                throw new PromotionException( "Execution of path promotion resume failed.", e );
-            }
-        }
-    }
-
-    private Future<PathsPromoteResult> submitResumePathsPromote( PathsPromoteResult result, String baseUrl )
-            throws IndyWorkflowException
+                    throws IndyWorkflowException
     {
         return detectOverload( () -> asyncPromotionService.submit( () -> {
             PathsPromoteResult ret;
             try
             {
-                ret = doResumePathsPromote( result, baseUrl );
+                ret = doPathsPromotion( request, false, baseUrl );
```

Comment:
Before we start the promotion process, can we make sure we've generated a promotion ID and set it in the MDC? This would probably help us debug promotion problems. If we can return that generated promotion ID to the caller, we can probably establish a search key for our logs.

Even better would be allowing them to (optionally) specify a promotion ID in the request, then generate one if it's not specified. That would allow the caller to specify the search key for logs related to this promotion.

Reply:
+1. We generate a promotionID in AbstractPromoteResult and return to caller. But I think we put it in the wrong place. As you mentioned, it should be in AbstractPromoteRequest. I updated this.